feat: HEALTH_READY.md documenting /health and /ready (#27477)
influxdata/influxdb: HEALTH_READY.md

- **`/health`** — liveness contributors. Reports whether each subsystem
  is currently healthy. Recomputed on every request. Use it for ongoing
  liveness monitoring after startup has completed.

Both endpoints are mounted by `HealthReadyHandler` at root **before** the
full API handler is installed (`http/check_handler.go`,
`http/api_handler.go`). They are available from the moment the PID file
is written. Until the main API handler is installed, any request to a
path other than `/health` or `/ready` receives `503 Service Unavailable`
with the body:

```json
{"status":"starting"}
```

This pre-API 503 is byte-exact (pinned by
`http/check_handler_jsoncompat_test.go`).

## Endpoint summary

| Path                            | Methods | 200 body status | 503 body status   |
|---------------------------------|---------|-----------------|-------------------|
| `/health`, `/health/`           | GET     | `"pass"`        | `"fail"`          |
| `/ready`, `/ready/`             | GET     | `"ready"`       | `"starting"`      |
| any other path (pre-API)        | any     | —               | `"starting"`      |

Trailing slashes are accepted on both endpoints. Query parameters are

> ignored — `/health?cachebust=1` is matched as `/health`.

Reviewer: ````suggestion
ignored--for example, `/health?cachebust=1` is matched as `/health`.
````

Path constants are defined in `http/handler.go` as `HealthPath` and
`ReadyPath`.

## Response headers

Every response written by `HealthReadyHandler` carries:

| Header              | Value                              |
|---------------------|------------------------------------|
| `Content-Type`      | `application/json; charset=utf-8`  |
| `X-Influxdb-Build`  | `OSS`                              |
| `X-Influxdb-Version`| The build version string           |

Source: `http/check_handler.go`, `NewHealthReadyHandler`.

---

## `/health`

### JSON envelope

> ```

Reviewer: Add a language identifier. This envelope isn't valid JSON (it uses `//` comments and `"pass" | "fail"` type unions), so `text` is the right fence.

````suggestion
```text
````

{
  "name":    "influxdb",          // constant
  "status":  "pass" | "fail",     // aggregate over all health checks
  "message": "<top-level message>",
  "checks":  [ <Check>, ... ],    // always present; may be []
  "version": "<build version>",
  "commit":  "<build commit>"
}
```

Each entry in `checks` is:

> ```

Reviewer: Per-check envelope schematic, not valid JSON (`//` comments) — use `text`.

````suggestion
```text
````

      "status": "fail",
      "message": "unreachable"
    }
  ],
  "version": "<build version>",
  "commit": "<build commit>"
}
```

(JSON trees taken verbatim from `http/check_handler_jsoncompat_test.go`.)

### Registered `/health` checks

The launcher registers the following named health checks. The exact set
depends on whether the server is running with on-disk metadata or
in-memory metadata (test mode).

| Check name        | Disk mode | Memory mode |
|-------------------|:---------:|:-----------:|
| `bolt`            |     ✓     |             |
| `sqlite`          |     ✓     |      ✓      |
| `query`           |     ✓     |      ✓      |
| `influxql`        |     ✓     |      ✓      |
| `task-scheduler`  |     ✓     |      ✓      |
| `shards`          |     ✓     |      ✓      |

Source of truth: `cmd/influxd/launcher/health_ready_test.go`.

#### `bolt`

Background prober that runs a no-op `bolt.View` every
`DefaultProbeInterval` (1 second). The result is wrapped in a
`FreshnessResponse` with a staleness budget of `DefaultProbeStaleness`
(5 seconds). If no fresh probe result is recorded within the budget, the
check flips to `"fail"` with a message of the form:

> ```

Reviewer: Message-format template, not JSON — use `text`.

````suggestion
```text
````

requests.

Source: `query/bridges.go`, `influxql/query/proxy_executor.go`,
`cmd/influxd/launcher/launcher.go`.

#### `task-scheduler`

Compares the task scheduler's next-run timestamp (`TreeScheduler.When()`)
to wall-clock time. Possible states:

| Condition                                              | Status | Message                          |
|--------------------------------------------------------|--------|----------------------------------|
| `When()` returns zero (no scheduled work)              | pass   | `scheduler idle: no scheduled runs` |
| `When()` is in the future                              | pass   | `next run in <duration>`         |
| `When()` is in the past by ≤ 30s (the pulse threshold) | pass   | `on time, dispatch lag <duration>` |
| `When()` is in the past by > 30s                       | fail   | `scheduler stalled: next run due <duration> ago` |

The threshold is `DefaultSchedulerPulseThreshold = 30 * time.Second`
(`cmd/influxd/run/scheduler_pulse.go`).

#### `shards`

Reports the accumulated shard-load errors observed during engine
startup. `"pass"` until at least one shard fails to load; thereafter
`"fail"` with a message of the form:

> ```

Reviewer: Message-format template, not JSON — use `text`.

````suggestion
```text
````

<n> shard(s) failed to load: shard <id>: <err>; shard <id>: <err>; ...
```

The same name (`shards`) appears in `/ready` with different semantics —
see below.

Source: `cmd/influxd/run/startup_logger.go`.

---

## `/ready`

### JSON envelope

> ```

Reviewer: `/ready` envelope schematic (`//` comments, `|` unions), not valid JSON — use `text`.

````suggestion
```text
````

  not change across requests.
- `up` is `time.Since(started)`, formatted as a `toml.Duration` string
  (e.g. `"2.5s"`, `"1h23m45.6s"`).
- `checks` is `omitempty`. On a `200` response it is **absent
  entirely**, not an empty array. On a `503` it contains **only the
  failing** gates.

Source: `http/check_handler.go`.

### Status codes

- `200 OK` with body `"status": "ready"` when **every** registered ready
  gate has been signaled `Ready()`.
- `503 Service Unavailable` with body `"status": "starting"` when **any**
  gate has not yet been signaled (or has been signaled `Unready()` during
  shutdown).

### Example: 200 (no failing checks)

```json
{
  "status": "ready",
  "started": "2026-05-26T15:42:30.123456789Z",
  "up": "1m32.4s"
}
```

`checks` is omitted by `omitempty`.

### Example: 503 (one gate unsignaled)

```json
{
  "status": "starting",
  "started": "2026-05-26T15:42:30.123456789Z",
  "up": "1.2s",
  "checks": [
    {
      "name": "metastores",
      "status": "fail",
      "message": "not ready"
    }
  ]
}
```

The literal message `"not ready"` is the default that every unsignaled
`ReadyGate` emits (`kit/check/helpers.go`). The `shards` gate emits

> different messages — see below.

Reviewer: Lead with the actor (active voice) and split the trailing clause.

````suggestion
Every unsignaled `ReadyGate` emits the default message `"not ready"`
(`kit/check/helpers.go`). The `shards` gate emits different messages.
See below.
````

### Registered ready gates

The launcher registers eight ready gates in this order:

1. `bolt` — KV (BoltDB) migrations complete.
2. `sqlite` — SQLite migrations complete.
3. `engine` — storage engine `Open` returned successfully.
4. `replications` — replication service `Open` returned successfully.
5. `query` — Flux query controller initialized.
6. `tasks` — task system initialized (pre-signaled when started with
   `--no-tasks`).
7. `task-scheduler` — `TreeScheduler` started.
8. `shards` — shard enumeration and loading complete (see below for
   progress states).

Source of truth: `cmd/influxd/launcher/health_ready_test.go` and
`cmd/influxd/launcher/subsystems.go`.

Each gate is binary: a single `Ready()` call latches it to `"pass"` for
the life of the process. During shutdown, the launcher walks the
labeled-closer chain and calls `Unready()` on every gate it owns, so

> `/ready` flips back to `503` while subsystems are being shut down.

Reviewer: Plain style: split the sentence and make the shutdown clause active. Kept "latch" since it pairs with the gate metaphor.

````suggestion
Each gate is binary. A single `Ready()` call latches it to `"pass"` for
the life of the process. During shutdown, the launcher calls `Unready()`
on every gate it owns, so `/ready` returns `503` while InfluxDB shuts
down those subsystems.
````

other than `/health` or `/ready` returns `503` with:

```json
{"status":"starting"}
```

This includes `/api/v2/*`, `/query`, `/write`, etc. Once the main API
handler is installed via `SetHandler`, those paths are served normally.

### Picking the right endpoint

Use `/ready`:

- To gate traffic during boot — for example, in a load balancer's
  health check or a Kubernetes startup probe.
- To observe startup progress (the `shards` gate reports a percentage).
- For a clean shutdown signal: when the process begins shutting down,
  the launcher flips gates back to `Unready()` and `/ready` returns
  `503` before the listener stops accepting connections.

Use `/health`:

- For ongoing liveness checks once the process is up.
- To detect transient degradation (a stuck `bolt` probe, a stalled
  scheduler, a shard load failure that occurred after startup
  completed).

Key behavioral difference: once a `/ready` gate has been signaled
`Ready()`, it does not transition back to `"fail"` unless the launcher
calls `Unready()` (shutdown). So `/ready` will not flap during normal
operation. `/health` is recomputed on every request and **will** reflect

> transient failures.

Reviewer: This is the key operational takeaway — wrap it in an IMPORTANT callout so it stands out.

````suggestion
> [!IMPORTANT]
> Once a `/ready` gate has been signaled `Ready()`, it does not transition
> back to `"fail"` unless the launcher calls `Unready()` (shutdown). So
> `/ready` will not flap during normal operation. `/health` is recomputed
> on every request and **will** reflect transient failures.
````

### curl examples

Probe `/ready` and `/health` and inspect the body:

```
$ curl -sS -o body.json -w '%{http_code}\n' http://localhost:8086/ready
200
$ jq . body.json
{
  "status": "ready",
  "started": "2026-05-26T15:42:30.123456789Z",
  "up": "1m32.4s"
}

$ curl -sS -o body.json -w '%{http_code}\n' http://localhost:8086/health
200
$ jq . body.json
{
  "name": "influxdb",
  "status": "pass",
  "message": "healthy",
  "checks": [
    { "name": "bolt", "status": "pass" },
    { "name": "sqlite", "status": "pass" },
    { "name": "query", "status": "pass" },
    { "name": "influxql", "status": "pass" },
    { "name": "task-scheduler", "status": "pass", "message": "scheduler idle: no scheduled runs" },
    { "name": "shards", "status": "pass" }
  ],
  "version": "<build version>",
  "commit": "<build commit>"
}

> ```

Reviewer: Drop the `$` prompt so the command copies cleanly, and split request from response so each block is single-purpose.

````suggestion
Probe `/ready` and inspect the response--for example:

```sh
curl -sS -o body.json -w '%{http_code}\n' http://localhost:8086/ready
```

The command prints the HTTP status code to stdout and writes the
response body to `body.json`. A ready instance returns `200` and the
following body:

```json
{
  "status": "ready",
  "started": "2026-05-26T15:42:30.123456789Z",
  "up": "1m32.4s"
}
```

Probe `/health` the same way:

```sh
curl -sS -o body.json -w '%{http_code}\n' http://localhost:8086/health
```

A healthy instance returns `200` and a body that lists each check:

```json
{
  "name": "influxdb",
  "status": "pass",
  "message": "healthy",
  "checks": [
    { "name": "bolt", "status": "pass" },
    { "name": "sqlite", "status": "pass" },
    { "name": "query", "status": "pass" },
    { "name": "influxql", "status": "pass" },
    { "name": "task-scheduler", "status": "pass", "message": "scheduler idle: no scheduled runs" },
    { "name": "shards", "status": "pass" }
  ],
  "version": "<build version>",
  "commit": "<build commit>"
}
```
````

During startup, `/ready` typically shows the `shards` gate progressing:

```
$ curl -sS -o body.json -w '%{http_code}\n' http://localhost:8086/ready
503
$ jq . body.json
{
  "status": "starting",
  "started": "2026-05-26T15:42:30.123456789Z",
  "up": "4.2s",
  "checks": [
    { "name": "engine",       "status": "fail", "message": "not ready" },
    { "name": "replications", "status": "fail", "message": "not ready" },
    { "name": "query",        "status": "fail", "message": "not ready" },
    { "name": "tasks",        "status": "fail", "message": "not ready" },
    { "name": "task-scheduler","status": "fail", "message": "not ready" },
    { "name": "shards",       "status": "fail", "message": "loading shards 47.0% (94 / 200)" }
  ]
}

> ```

Reviewer: ````suggestion
Probe `/ready` during startup to watch the `shards` gate progress--for example:

```sh
curl -sS -o body.json -w '%{http_code}\n' http://localhost:8086/ready
```

The command returns `503`, and the body lists only the checks that
haven't passed:

```json
{
  "status": "starting",
  "started": "2026-05-26T15:42:30.123456789Z",
  "up": "4.2s",
  "checks": [
    { "name": "engine",       "status": "fail", "message": "not ready" },
    { "name": "replications", "status": "fail", "message": "not ready" },
    { "name": "query",        "status": "fail", "message": "not ready" },
    { "name": "tasks",        "status": "fail", "message": "not ready" },
    { "name": "task-scheduler","status": "fail", "message": "not ready" },
    { "name": "shards",       "status": "fail", "message": "loading shards 47.0% (94 / 200)" }
  ]
}
```
````

Distinguishing a 200 response from a 503 from the body alone:

```
$ curl -sS http://localhost:8086/ready  | jq -r .status
ready

$ curl -sS http://localhost:8086/health | jq -r .status
pass

> ```

Reviewer: ````suggestion
Distinguish a `200` from a `503` using the body alone--for example:

```sh
curl -sS http://localhost:8086/ready | jq -r .status
```

```text
ready
```

Do the same for `/health`:

```sh
curl -sS http://localhost:8086/health | jq -r .status
```

```text
pass
```
````

The HTTP status code is the authoritative signal; the body string is for
human readability.

---

## Troubleshooting: per-subsystem failure scenarios

For each scenario below, the JSON snippet is the relevant portion of the
`/health` or `/ready` response — other fields elided for brevity.

### `/ready` 503 — shards still loading

```json
{
  "status": "starting",
  "checks": [
    { "name": "shards", "status": "fail", "message": "loading shards 47.0% (94 / 200)" }
  ]
}
```

**Meaning:** Normal startup. The engine has enumerated 200 shards and 94
have finished opening. Wait — the percentage updates each time another
shard completes. If the percentage stops climbing, check `influxd` logs
for `Finished loading shard` lines; whichever shard is currently in

> flight is logged when it completes.

Reviewer: Split "what it means" from "what to do." "Wait" here is an instruction, so make it an explicit imperative under Action.

````suggestion
**Meaning:** Normal startup. The engine has enumerated 200 shards and 94
have finished opening. The percentage updates each time another shard
completes.

**Action:** Wait for loading to finish. If the percentage stops climbing,
check `influxd` logs for `Finished loading shard` lines. The shard
currently in flight is logged when it completes.
````

### `/ready` 503 — shards enumeration not yet started

```json
{
  "status": "starting",
  "checks": [
    { "name": "shards", "status": "fail", "message": "waiting for shard enumeration" }
  ]
}
```

**Meaning:** The engine has not yet begun calling `AddShard`. Either the
engine is still in early initialization (look for the `engine` gate also
failing with `"not ready"`) or the engine is blocked before

> enumeration. Check logs for engine open progress.

Reviewer: ````suggestion
**Meaning:** The engine hasn't yet begun calling `AddShard`. The engine
is either still in early initialization or blocked before enumeration.

**Action:** Look for the `engine` gate also failing with `"not ready"`,
which indicates early initialization. Check logs for engine open progress.
````

### `/ready` 503 — terminal shard load failure

```json
{
  "status": "starting",
  "checks": [
    { "name": "shards", "status": "fail", "message": "shard loading failed: open /var/lib/influxdb2/engine/data/2/.../000000001-000000001.tsm: input/output error" }
  ]
}
```

**Meaning:** `engine.Open` returned an error and the `shards` gate has
latched into a terminal failure. Restarting will not clear this without
addressing the underlying error in the message. Check disk health and

> file permissions on the data directory.

Reviewer: ````suggestion
**Meaning:** `engine.Open` returned an error and the `shards` gate
latched into a terminal failure. Restarting won't clear this without
addressing the underlying error in the message.

**Action:** Check disk health and file permissions on the data directory.
Address the underlying error before you restart.
````

### `/health` 503 — bolt prober stale

```json
{
  "status": "fail",
  "message": "stale: last probe 12.3s ago (threshold 5s)",
  "checks": [
    { "name": "bolt", "status": "fail", "message": "stale: last probe 12.3s ago (threshold 5s)" }
  ]
}
```

**Meaning:** The background bolt prober has not recorded a fresh result
within the 5-second staleness budget. Either the prober goroutine is
wedged in a `db.View` (the bolt mmap is unresponsive — check disk and
kernel logs) or the host is under such severe scheduling pressure that
the 1-second probe loop cannot run. Look at `iostat`, `dmesg`, and the

> `influxd` process's CPU/memory state.

Reviewer: ````suggestion
**Meaning:** The background bolt prober hasn't recorded a fresh result
within the 5-second staleness budget. Either the prober goroutine is
wedged in a `db.View` and the bolt mmap is unresponsive, or the host is
under severe scheduling pressure that prevents the 1-second probe loop
from running.

**Action:** Check disk and kernel logs for an unresponsive mmap. Inspect
`iostat`, `dmesg`, and the `influxd` process's CPU and memory state.
````

### `/health` 503 — scheduler stalled

```json
{
  "status": "fail",
  "message": "scheduler stalled: next run due 1m45s ago",
  "checks": [
    { "name": "task-scheduler", "status": "fail", "message": "scheduler stalled: next run due 1m45s ago" }
  ]
}
```

**Meaning:** The task scheduler's next-run timestamp is more than 30
seconds behind wall clock — its dispatch loop fired a timer but never
advanced. Look for blocked goroutines in the scheduler (use the runtime
profile or stack dump). If this fires only at boot and clears within a

> few seconds, it is likely a cold-start dispatch and not a real wedge.

Reviewer: ````suggestion
**Meaning:** The task scheduler's next-run timestamp is more than 30
seconds behind wall clock. Its dispatch loop fired a timer but never
advanced. If this fires only at boot and clears within a few seconds,
it's likely a cold-start dispatch and not a real wedge.

**Action:** Look for blocked goroutines in the scheduler using the
runtime profile or a stack dump.
````

### `/health` 503 — sqlite not open

```json
{
  "status": "fail",
  "message": "sqlite database not open",
  "checks": [
    { "name": "sqlite", "status": "fail", "message": "sqlite database not open" }
  ]
}
```

**Meaning:** The SQLite metadata store handle is nil — the store was
closed or never opened. This should only be observed transiently during
startup (before SQL migrations complete) or during shutdown. If it
persists, the SQLite store failed to open; check `influxd` logs for the

> underlying error.

Reviewer: Remove the passive "should be observed" and name the actor/timing.

````suggestion
**Meaning:** The SQLite metadata store handle is nil because the store
was closed or never opened. InfluxDB reports this state only briefly,
while it starts up (before SQL migrations complete) or while it shuts
down.

**Action:** If the failure persists, the SQLite store failed to open.
Check `influxd` logs for the underlying error.
````

### `/health` 503 — sqlite ping error

```json
{
  "status": "fail",
  "message": "context deadline exceeded",
  "checks": [
    { "name": "sqlite", "status": "fail", "message": "context deadline exceeded" }
  ]
}
```

**Meaning:** `db.PingContext` did not return within the 500ms probe
deadline, or returned an I/O error. Check disk health on the SQLite
database file and whether another process is holding a long-running

> write transaction.

Reviewer: ````suggestion
**Meaning:** `db.PingContext` didn't return within the 500ms probe
deadline, or it returned an I/O error.

**Action:** Check disk health on the SQLite database file and whether
another process is holding a long-running write transaction.
````

### `/health` 503 — accumulated shard load failures

```json
{
  "status": "fail",
  "message": "3 shard(s) failed to load: shard 41: I/O error; shard 87: corrupt index; shard 102: I/O error",
  "checks": [
    { "name": "shards", "status": "fail",
      "message": "3 shard(s) failed to load: shard 41: I/O error; shard 87: corrupt index; shard 102: I/O error" }
  ]
}
```

**Meaning:** One or more shards failed to load during engine open.
Errors accumulate — even if the engine itself opened successfully,
`/health` will continue to report this until restart. Each `shard <id>`
in the message identifies a specific shard directory under the
configured data path; address each shard's underlying error before

> restarting.

Reviewer: ````suggestion
**Meaning:** One or more shards failed to load during engine open. Errors
accumulate. Even if the engine opened successfully, `/health` continues
to report this until restart.

**Action:** Each `shard <id>` in the message identifies a specific shard
directory under the configured data path. Address each shard's underlying
error before you restart.
````

### `/health` 503 — all checks pass but the body says `"fail"`

If `/health` returns `503` but the `checks` array contains only
`"pass"` entries, the response is being constructed under an
inconsistent observation (a check transitioned between the aggregation
walk and the per-check render). Retry the request — the inconsistency
window is bounded by the probe interval. If the condition persists

> across many requests, file an issue with the body and headers attached.

Reviewer: This scenario had no **Meaning:** label — added the Meaning/Action structure so all scenarios are parallel.

````suggestion
**Meaning:** `/health` returned `503` but the `checks` array contains
only `"pass"` entries. The response was constructed under an inconsistent
observation, where a check transitioned between the aggregation walk and
the per-check render.

**Action:** Retry the request. The inconsistency window is bounded by the
probe interval. If the condition persists across many requests, file an
issue with the body and headers attached.
````
